### DISCUSSION TITLE
Draft Snippets functionality

### DISCUSSION THREAD

**Discussion by phildexter** [wagtail/wagtail#8599] [Ideas]
Hello Wagtail 👋 

We (@kaedroho @laymonage and I) are making some improvements to Snippets functionality, which includes introducing the concept of publish and draft status for Snippets, just like for Pages. We have a question about what it means for a Snippet only to have a draft. We also have a proposed solution that we're interested in people's thought on. 

To explain this scenario, it's helpful to think about how Wagtail currently handles it Pages. 

Let's say you create a new instance of a `Blog Page`, where you have a page chooser to chose an `Author page`. That page chooser will let you select any `Author page`, regardless of status - so you can easily choose a draft `Author page` that doesn't have a published version. If you publish the blog, any author data that is shown on the blog page template is pulled through to the frontend, even though the author page itself isn't published. This scenario feels wrong, but it's also not the specific use-case we're trying to solve here, it's just highlighting existing behaviour that (I think) we should look to fix in the future.

Now imagine the scenario above, but instead of the `Author page` you have a Snippet called `Author`, which (in our brave new world) is in draft status. Should we copy the exact same functionality that happens for pages right now - even though we think this might not be optimal? Or should we do something different?

Having thought about this quite hard, here are where my thoughts (heavily influenced by many other people) are at.

1. We shouldn't copy what happens to pages. The fact this happens with pages feels strange - it feels like we shouldn't show draft information on a live page. With Snippets this problem is likely to happen more often, as editors retrieve info from Snippets more often than they do from pages. More strangeness isn't good.

2. Ultimately, we should use the (yet to be defined) [object usage functionality](https://github.com/wagtail/wagtail/discussions/8496) to help with this. This could prevent editors from publishing pages that have links to draft snippets/pages. It could also stop editors from unpublishing snippets/pages that are linked to from live pages. However, this is far from being implemented and we'd like to be able to release the new Snippets functionality before the object usage tech and UX is fully defined and released.

3. So, if we don't want to copy pages, and we don't have object usage, what is the next best thing we can do right now. I believe we should ensure there is always a published version of a Snippet instance. This would mean the first time you create a Snippet, your only option is to publish it. After that you could benefit from draft and publish statuses. You also wouldn't be able to unpublish a snippet, only delete it. This is very close to how Snippets work now, so shouldn't make things worse - whilst also bringing the benefits of drafts (which also allows previews) to existing Snippets.

To me, this feels like a sensible compromise from a technical and UX perspective. Keen to hear thoughts.

**Comment by Scotchester** [comment]
I think that is a solid proposal, @phildexter, but I want to pose another option:

If we were to allow snippets to be initially created as unpublished drafts, could we make it so that the snippet chooser will only allow choosing snippets that have been published?

**Reply by phildexter** [reply]
We could do this yeah 😄 

My concerns with this approach are mostly around editors getting confused about Wagtail's logic. There would be 2 changes that could be quite confusing.

1. This behaviour is different from pages. So when we bring drafts in, users will need to figure out why only some are showing, when for pages all show. (If we also changed the page chooser to follow suit, it would stop people from preparing a set of pages that relate and link to each other, which I think isn't good.)
2. When we move onto the next stage (object usage and warning editors) we would then want to allow draft snippets to be chosen again (so that a set of draft snippets and page could be prepared for a big go live). At that point we're flipping from the behaviour we had in version 1, to a new version.

I'm unsure if the benefits outweigh the costs of that approach....maybe they do?

**Reply by Scotchester** [reply]
Great points, for sure.

I also want to connect this with https://github.com/wagtail/wagtail/discussions/8496, in which Karl's original post states:

> We think that when a snippet is in draft state, it would not be selectable in any snippet choosers.

Does this represent a change in that original thinking?

**Reply by phildexter** [reply]
Yep, we should probably update that @kaedroho as our thinking has evolved since writing that...

I'm certainly not writing it off, just trying to weigh up which is the least bad option 😄

**Comment by NotANormalNerd** [comment]
So just to give our input:
TLDR: Don't let people choose unpublished snippets, warn if they unpublish a used snippet, let snippets have a draft status.

Currently we have to use `Choose a Page` workaround to build a multi part teaser, that would be better suited with Snippets.

We have the same problem and found the following solution to the problem: On the `TeaserPage` you can only choose live `PanelPages`, this is easily implemented in the chooser by a Queryset. When a editor unpublishes the `PanelPage` we assume it is for a very good reason and just try to degrade gracefully, but will remove the `PanelPage` from the `TeaserPage`.

It would be nice to have a popup that will show at least: "This snippet is used on the following pages: Are you sure?", but we just train the editors on that and additionally added a helptext so they know of that behaviour.

**Reply by phildexter** [reply]
> Don't let people choose unpublished snippets, warn if they unpublish a used snippet, let snippets have a draft status.

Yep I totally agree. This kind of thing is where we'd like to get to, but I don't think we'll have it for Wagtail 4 (August), so what I'm proposing is a temporary solution until we can implement something more comprehensive.

**Reply by NotANormalNerd** [reply]
Also: Why not just give developers a proper API to choose if they want live, draft or scheduled pages?

**Reply by phildexter** [reply]
Errr I don't know. I guess this discussion is focusing on what it means for a Snippet to be in draft. Is your last point related to this @NotANormalNerd or are you just unsure where to raise this?

**Comment by th3hamm0r** [comment]
I think I have a different view based our projects, but first: thanks for your awesome work on those improvements!

My view:
- similar to pages, allow choosing unpublished snippets, allow creation of initially unpublished snippets (there are some options for the developer to avoid those cases, if required, e.g. model validation, django's messaging framework)
- instead of changing the behavior, we should try to improve the UI of the choosers, showing the publish-state of the page/snippet similar to the page's list views (e.g. "live", "draft", "scheduled",...)
- further we could provide options for the developer to filter the selectable pages/snippets (e.g. based on the publish-state)


> Let's say you create a new instance of a `Blog Page`, where you have a page chooser to chose an `Author page`. That page chooser will let you select any `Author page`, regardless of status - so you can easily choose a draft `Author page` that doesn't have a published version. If you publish the blog, any author data that is shown on the blog page template is pulled through to the frontend, even though the author page itself isn't published.

In our past projects we rarely had those issues, because at the end it is up to the developer to filter the (related) objects before displaying it (e.g. by using checks like `{% if page.author_page.live %}...display...{% endif %}`).

Especially with scheduled publishing, creating multiple draft objects before publishing them at a specific time is quite important for us.

For example, when we use `StreamField`s to manage pinned teaser items, those items often are only published for a scheduled timeframe. In that case, the editor needs to pin pages (or snippets in the near future), which aren't published yet.
Here it would be really helpful, if the choosers would display more information on the draft state. We sometimes helped us with custom outputs in the CMS, but having that out-of-the box in the choosers would be nice.

>     3. So, if we don't want to copy pages, and we don't have object usage, what is the next best thing we can do right now. I believe we should ensure there is always a published version of a Snippet instance. This would mean the first time you create a Snippet, your only option is to publish it. After that you could benefit from draft and publish statuses. You also wouldn't be able to unpublish a snippet, only delete it. This is very close to how Snippets work now, so shouldn't make things worse - whilst also bringing the benefits of drafts (which also allows previews) to existing Snippets.

As above, this should be allowed, in my view, the developer should be able to handle those cases (gracefully).
You could still provide (optional) usage information, which may help in other cases, but disallowing snippets to be unpublished completely seems to be very restrictive. You don't always want to delete an object, just to hide it from the frontend (temporarily) ;) 

My view is probably influenced by the heavy usage of the [SilverStripe CMS](https://www.silverstripe.org), before we mainly moved to wagtail/django. SilverStripe for example, based on its existing draft-handling (used for assets, custom objects, pages,...), has a "campaign" feature, where editors can kind of orchestrate the publishing process of many objects with a single click. So you could prepare everything in draft-mode and publish it at the end with such a campaign.

**Reply by th3hamm0r** [reply]
After reading my post again, a summary in other words 😅:
I actually like the way how it is currently implemented with pages and being able to choose draft pages (so I would be happy to have that for draft snippets too), BUT there are cases where the UI could really display more information to help the editor understanding the output (eg. publish-state of the page for choosers).

**Reply by phildexter** [reply]
Thanks for the feedback. 

I actually think I agree with most of what you are saying - specifically the point about draft snippets being able to be selected (which I recognise is contrary to what I wrote [here](https://github.com/wagtail/wagtail/discussions/8599#discussioncomment-2828241).)

I think longer term I'd like:
 - Draft snippets to be create-able and selectable
 - Better UI to help editors know what is draft vs live when choosing objects
 - Better help for editors when they try and publish and object that references another object that is not published (not stopping them from doing it, but warning them / helping them)\

However, in the very short term we want to release something for Wagtail 4 to get some of the benefits, and we won't be able to get the three bullets above done. Therefore, I'm thinking that interim state (that we release in Wagtail 4) should still be point 3 in my original comment...

Additionally: I like the idea of a campaign, where you can set a date and time and select multiple things that are in draft to be published at once. Obviously out of scope of this, but it would be a nice thing to introduce independently at some point.

**Reply by th3hamm0r** [reply]
The long term plan sounds good to me, and I totally understand your point for the short term!
I'm not completely sure, if copying the page's implementation would actually be easier to implement, and it's more the direction of the long-term plan.
At the end the whole functionality is optional, so no matter which option of those you will choose, nobody gets hurt and some people will still be able to benefit from it 👍

**Reply by phildexter** [reply]
> At the end the whole functionality is optional, so no matter which option of those you will choose, nobody gets hurt and some people will still be able to benefit from it 👍

Good point thanks! Easy to forget that people have to take action to instigate this rather than it being pushed upon them.

**Comment by ababic** [comment]
> We shouldn't copy what happens to pages. The fact this happens with pages feels strange - it feels like we shouldn't show draft information on a live page. With Snippets this problem is likely to happen more often, as editors retrieve info from Snippets more often than they do from pages. More strangeness isn't good.

From reading through the rest of the discussion, it feels like the position on this has changed slightly. But, in case I have misunderstood, I just wanted to share my thoughts on this from a developer perspective.

From a 'product' perspective, I can totally see why you would take this view; But, SilverStripe and other CMS platforms have a stronger involvement in the rendering process, where this kind of 'magic' can be applied. Essentially, all of the 'dynamic' stuff is handled by the platform. 

Self-hosted Django projects are quite different. Wagtail might be used for anything from the entire project to only small parts of it, and practically any part of the behaviour can be overridden or patched in some way. With this in mind, it makes far less sense for Wagtail to do anything too 'magic' when it comes to draft handling. To control which objects are returned would require 'hacking' the model managers to always apply filters under certain circumstances, which feels very wrong to me; What if, as a developer, you need to access ALL items? How would a preview work? Would that bypass the filtering somehow? The much more 'Django' option is to offer developers transparent `.live()` and `.not_live()` filters (like we do for pages) and encourage them to apply these where relevant.

It's worth saying though: If we are talking about Wagtail as a hosted service/platform (where developers are locked out of the underlying code), that is a different story all together :)
